scenarios/common/infrastructure_deployment.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2019 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+import os
+import time
+import uuid
+import shutil
+
+from python_terraform import Terraform
+
+execution_id = str(uuid.uuid1())[:8]
+
+
+def is_infra_deployment_node(config):
+    if hasattr(config, 'slaveinput'):
+        return config.slaveinput.get('slaveid') == 'gw0'
+    return True
+
+
+def infrastructure_deployed(deployment_hash, success=True):
+    filename = os.path.join(os.path.dirname(__file__), deployment_hash)
+    with open(filename, 'wt') as f:
+        if success:
+            f.write(execution_id)
+        else:
+            f.write('Failed')
+
+
+def tests_finished(deployment_hash, id):
+    filename = os.path.join(os.path.dirname(__file__), deployment_hash + '.' + id)
+    with open(filename, 'wt') as f:
+        f.write('Done')
+
+
+def wait_for_infrastructure(deployment_hash):
+    filename = os.path.join(os.path.dirname(__file__), deployment_hash)
+    while True:
+        if os.path.exists(filename):
+            with open(filename, 'rt') as f:
+                global execution_id
+                execution_id = f.read()
+                if execution_id == 'Failed':
+                    return False
+                return True
+        else:
+            time.sleep(5)
+
+
+def wait_for_tests_finished(config, deployment_hash):
+    done_files = []
+    filename = os.path.join(os.path.dirname(__file__), deployment_hash)
+
+    if hasattr(config, 'slaveinput'):
+        done_files = [filename + '.gw' + str(i)
+                      for i in range(1, config.slaveinput['workercount'])]
+
+    while True:
+        if all(os.path.exists(file) for file in done_files):
+            break
+        else:
+            time.sleep(5)
+
+    os.remove(filename)
+    for file in done_files:
+        os.remove(file)
+
+
+def terraform_apply(working_dir, deployment_hash):
+    terraform = Terraform(working_dir=working_dir)
+    return_code, stdout, stderr = terraform.init()

```

Comment:
doing terraform init every apply feels a bit wasteful, its a lot of redownloading the same tf providers over and over.

Reply:
At the moment `terraform.apply` runs once per execution.

If we decide to run it multiple times, `terraform init` is safe to run multiple times and it is a no-op if there are no changes in the providers.

Reply:
tf init is local to the work directory, given different providers, it would make sense to have a global init for the providers (either all or per needed set) and copy them in for specific executions. ie. yes its safe, but its slow to re-run for each directory.

Reply:
Yeah, I misunderstood your first comment. Updated it now to initialize providers in `~\.cloud-custodian\terraform` folder & use it for `terraform init`